stellargraph/layer/ppnp.py

```diff
@@ -0,0 +1,188 @@
+
+      - The size of the final fully connected layer must be equal to the number of classes
+
+    Args:
+        layer_sizes (list of int): list of output sizes of fully connected layers in the stack
+        activations (list of str): list of activations applied to each fully connected layer's output
+        generator (FullBatchNodeGenerator): an instance of FullBatchNodeGenerator class constructed on the graph of interest
+        bias (bool): toggles an optional bias in fully connected layers
+        dropout (float): dropout rate applied to input features of each layer
+        kernel_regularizer (str): normalization applied to the kernels of fully connetcted layers
+        transport_probability: "probability" of returning to the starting node in the propogation step as desribed  in
+        the paper (alpha in the paper)
+    """
+
+    def __init__(
+            self,
+            layer_sizes,
+            activations,
+            generator,
+            bias=True,
+            dropout=0.0,
+            transport_probability=0.1,
+            kernel_regularizer=None
+    ):
+
+        if not isinstance(generator, FullBatchNodeGenerator):
+            raise TypeError("Generator should be a instance of FullBatchNodeGenerator")
+
+        assert len(layer_sizes) == len(activations)
+
+        self.layer_sizes = layer_sizes
+        self.transport_probability = transport_probability
+        self.activations = activations
+        self.bias = bias
+        self.dropout = dropout
+        self.kernel_regularizer = kernel_regularizer
+        self.generator = generator
+        self.support = 1
+        self.method = generator.method
+
+        # Check if the generator is producing a sparse matrix
+        self.use_sparse = generator.use_sparse
+        if self.method == "none":
+            self.graph_norm_layer = GraphPreProcessingLayer(
+                num_of_nodes=self.generator.Aadj.shape[0]
+            )
+
+        # Initialize a stack of fully connected layers
+        n_layers = len(self.layer_sizes)
+        self._layers = []
+        for ii in range(n_layers):
+            l = self.layer_sizes[ii]
+            a = self.activations[ii]
+            self._layers.append(Dropout(self.dropout))
+            self._layers.append(
+                Dense(
+                    l,
+                    activation=a,
+                    use_bias=self.bias,
+                    kernel_regularizer=self.kernel_regularizer,
+                )
+            )
+
+        # Create propogation matrix and propogation layer
+        Aadj = generator.Aadj.todense()
+
+        # propagation matrix is non-sparse : create and store at initalization to save time
+        propogation_matrix = self.transport_probability * np.linalg.inv(np.eye(Aadj.shape[0]) -
+                                                                  (1 - self.transport_probability) * Aadj)
+        propogation_matrix = tf.convert_to_tensor(propogation_matrix, tf.float32)
```

Comment:
Here you take the matrix in the generator and create a static tensor representation of the matrix. This means you cannot use the same model on different graphs as the model contains the graph. This is an important caveat of this approach: PPNP embeds the graph in the model so using a generator with another graph will fail, whereas APPNP as it uses the `GraphConvolutionalLayer` the model can be transferred between graphs by simply providing a generator object with the new graph.

Reply:
Rather than doing this I would use a graph multiplication layer, as for the APPNP code. We can avoid calculating the inverse in the graph each time the model is run, by adding an option to the `FullBatchNodeGenerator` to do this pre-processing. The `FullBatchNodeGenerator` could then be called with `method='ppnp'` instead of `method='gcn'`.

Reply:
A note for future development: having the pre-processed adjacency matrix as a static tensor that can be stored in GPU memory rather than transferred at each iteration of training is a good idea.  However, to do this and allow applying or training the model on multiple graphs, we need to find a way to allow updating this adjacency matrix when we change the graph that is easy for the user.

Reply:
@adocherty good idea, I'll add this into the FullBatchNodeGenerator and either add a warning or an error if a user tries to use `sparse=True` while using `method="ppnp"` so they don't accidentally run things expecting a sparse matrix.